torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
       "Work needs to be completed before calling result(). "
       "Should call wait() before result().");
   TORCH_CHECK(
-      outputTensors_.size() <= 1, "work result does not support list of lists.");
+      outputTensors_.size() <= 1,
+      "work result does not support list of lists, use .getFuture() and value()");
   return outputTensors_.size() == 0 ? std::vector<at::Tensor>()
                                     : outputTensors_.at(0);
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupGloo::AsyncWork::
+    getFuture() {
+  return future_;
+}
+
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> createFutureAsOutput(
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() > 1) {
+    return c10::make_intrusive<c10::ivalue::Future>(
+        c10::ListType::create(c10::ListType::create(c10::TensorType::get())));
+  }
+  return c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
```

Comment:
Is there a reason for this? I find it odd that a type information depends on a runtime condition. Types are typically useful to "predict what would happen before it happens" (e.g., static analysis) but by doing this we defeat that purpose.

Reply:
I'm not very knowledgeable on the PG API. My first impression is that it would be more "regular" and predictable if here we always used a list of tensors, even in case of a singleton.

If however you say that different methods of a PG need to return different types, that's fine, and some can return a list of tensors while other can return a singleton tensor. However I'd like if a certain method always returns the same type, no matter its arguments.

Reply:
agreed. I don't mind returning a singleton tensor wrapped in a list and we already do, but how would we deal with returning list of lists?